apps/aecore/src/aec_parent_connector.erl

```diff
@@ -326,3 +363,103 @@ get_sign_module() ->
+    gen_server:call(?SERVER, {create_pin_tx, SenderEnc, ReceiverPubkey, Amount, Fee, PinningData}).
+
+post_pin_tx(Tx) ->
+    gen_server:call(?SERVER, {post_pin_tx, Tx}).
+
+get_pin_by_tx_hash(TxHash) ->
+    gen_server:call(?SERVER, {get_pin_by_tx_hash, TxHash}).
+
+encode_parent_pin_payload(Pin) ->
+    gen_server:call(?SERVER, {encode_parent_pin_payload, Pin}).
+
+decode_parent_pin_payload(PinPayload) ->
+    gen_server:call(?SERVER, {decode_parent_pin_payload, PinPayload}).
+
+encode_child_pin_payload(TxHash) ->
+    gen_server:call(?SERVER, {encode_child_pin_payload, TxHash}).
+
+decode_child_pin_payload(TxHash) ->
+    gen_server:call(?SERVER, {decode_child_pin_payload, TxHash}).
+
+is_pin(Pin) -> 
+    gen_server:call(?SERVER, {is_pin, Pin}).
+
+% PINREFAC is this (aec_p_c) the correct place for this and the following CC pin related ones?
+-spec get_pinning_data() -> {ok, map()} | {error, atom()}.
+get_pinning_data() ->
+    {ok, #{epoch := Epoch,
+           first := First,
+           last  := Last}} = aec_chain_hc:epoch_info(),
+    lager:debug("Get pin data for epoch ~p for leader of block ~p", [Epoch - 1, Last]),
+    {ok, BlockHash} = aec_chain_state:get_key_block_hash_at_height(First-1),
+    {ok, ChainType} = get_parent_chain_type(),
+    PrevEpoch = Epoch - 1,
+    Height = First - 1,
+    case aec_consensus_hc:leader_for_height(Last) of
+      {ok, Leader} ->
+        {ok, #{epoch             => PrevEpoch,
+               height            => Height,
+               block_hash        => BlockHash,
+               parent_payload    => encode_parent_pin_payload(#{epoch => PrevEpoch, height => Height, block_hash => BlockHash}),
+               last_leader       => Leader,
+               parent_type       => ChainType,
+               parent_network_id => get_network_id()}};
+      error ->
+          %% schedule not yet cached
+          {error, last_leader_unknown}
+    end.
+
+get_generations(FromHeight, ToHeight) ->
+    ReversedBlocks =
+        lists:foldl(
+            fun(Height, Accum) ->
+                case aec_chain:get_generation_by_height(Height, backward) of
+                    {ok, #{micro_blocks := MBs}} ->
+                        ReversedGeneration = lists:reverse(MBs),
+                        ReversedGeneration ++ Accum;
+                    error -> error({failed_to_fetch_generation, Height})
+                end
+            end,
+            [],
+            lists:seq(FromHeight, ToHeight)),
+    lists:reverse(ReversedBlocks).
+
+
+find_spends_to(Account) ->
+    %% TODO Not happy with this. are we in the correct epoch when this is supposed to be 
+    %%      called, and what if you want to go back through the chain and validate backwards?
+    {ok, #{last := Last, first := First}} = aec_chain_hc:epoch_info(), 
+    Blocks = get_generations(First, Last-1),
+    lists:flatten([ pick_pin_spends_to(Account, Txs) || {mic_block, _, Txs, _} <- Blocks ]).
```

Comment:
Don't use the internal structures like `mic_block` there are accessor functions in `aec_blocks`

Reply:
Agreed. Can you point me to some place where this kind of thing is done in a better way, and I can mimic that style? These APIs are (at least not for the casual viewer) beacons of clarity and docs...

Reply:
Here you would need `aec_blocks:type/1` to do the filter and then `aec_blocks:txs/1` to access the transactions.